File: changhyun/백준문제/Boj13702.java
package 백준문제;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Boj13702 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int N, K;
    static int[] makgeollies;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        makgeollies = new int[N];

        int maxValue = Integer.MIN_VALUE;
        for (int i = 0; i < N; i++) {
            makgeollies[i] = Integer.parseInt(br.readLine());
            maxValue = Math.max(maxValue, makgeollies[i]);
        }


        long left = 1;
        long right = maxValue;

        while (left <= right) {

            long mid = (left + right) / 2;

            int cnt = 0;
            for (int makgeoli : makgeollies) {
                cnt += makgeoli / mid;
            }

            if (cnt >= K) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }

        }

        System.out.println(right);


    }
}

File: duna/Boj13702.py
n,k=map(int,input().split())
alcohol=[
  int(input()) for _ in range(n)
]


def can_drink(i):
  cnt=0
  for a in alcohol:
    cnt+=a//i
    #print(a,i)
  
  return cnt>=k

left=1
right=max(alcohol)

while left<=right:
  mid=(left+right)//2
  if can_drink(mid):
    ans=mid
    left=mid+1
  else:
    right=mid-1

print(ans)